### PR TITLE
Add sort description buttons

### LINKED ISSUE #188: Table sort buttons are not descriptive
## What

During accessibility review, it has been noted that usage of the `title` attribute to denote sorting action is not sufficient. Please see full comment below:

> Even though this has a title attribute advising of the button’s function, this is not advised to screen reader users as the title attribute normally isn’t read by screen reading software. Adding the information to the column header will cause the information to be read out every time a user crossed into a new column; so, it can be added to an aria-describedby attribute which be read after the label

## Details

A recommended solution is to add screen-reader accessible descriptions that can be placed into a visually hidden element, see example below:

```
<div style="display: none;">
    <span id="desc_SortCountry">Click to sort table by Country</span>
</div>
<button class="o-table__sort" title="sort table by Country" aria-descrivedby="desc_SortCountry">```

### PR DESCRIPTION
**Pull request by chee**
to close #188.

please try this out in whatever screenreaders you have available. i've tried it with Voiceover and Orca

### CODE REVIEW

File: src/js/Tables/BaseTable.js [hunk 1]
```diff
@@ -469,8 +469,16 @@ class BaseTable {
...
 			const nextSort = this._getNextSortOrder(th);
-			sortButton.setAttribute('title', `sort table by "${th.textContent}" ${nextSort}`);
+			const sortDescription = `sort table by "${th.textContent}" ${nextSort}`
+			const sortDescriptionElement = document.createElement("div")
+			sortDescriptionElement.style.display = 'none';
+			sortDescriptionElement.id = `sort-${Math.random().toString(36).slice(2)}`
+			sortDescriptionElement.textContent = sortDescription;
+			th.append(sortDescriptionElement);
+
+			sortButton.setAttribute("aria-describedby", sortDescriptionId);
```
Comment (JakeChampion) [inline src/js/Tables/BaseTable.js:481]: Should this be `sortDescriptionElement.id`?
`sortDescriptionId` doesn't exist, maybe some commits haven't been pushed to github

**Comment by JakeChampion** [comment]
As discussed in our meeting, here is the list of screenreaders gov uk tests with -- https://www.gov.uk/service-manual/technology/testing-with-assistive-technologies

We should test with those ones plus Orca, which is a screenreader available on linux

**Comment by origamiserviceuser** [comment]
o-table bundle size difference from 8.1.4 to 7cbd26a37d7399d5e571e3dce6cb64e3ad47dda1
js: 0.25kb increase (0.10kb/gzip)
An insignificant difference was also found for: css (master), css (internal), css (whitelabel)

**Comment by chee** [comment]
OK, i have tested this in Orca, JAWS, VoiceOver and NVDA and i haven't found an instance where it makes a difference which one of these two approaches we use.

This code and the current code seem to operate exactly the same in all of those screenreaders.

**Comment by chee** [comment]
> OK, i have tested this in Orca, JAWS, VoiceOver and NVDA and i haven't found an instance where it makes a difference which one of these two approaches we use.
> 
> This code and the current code seem to operate exactly the same with all of those screenreaders.

hoping somebody with more experience with screenreaders can check this out?

**Comment by notlee** [comment]
Copied from Slack, because Slack is not open and makes it harder to find history:


sorry I’ve only just spotted this! excellent work getting those assistive techs setup and trying it out in each of them :clap: it’s frustrating that the proposed changes doesn’t help, but that’s good to know. When I’m finished with the Build Service thing I’m going to put time aside to get these installed so I can be more helpful

there’s maybe not a perfect solution, the gds design system doesn’t seem to have a sortable table to look to although the [MOJ specifically does](https://github.com/alphagov/govuk-design-system-backlog/issues/61). it reads “up arrow [content] down arrow” in some scenarios which isn't ideal 😬 
Screenshot 2021-02-09 at 16 50 02
but interestingly that isn’t re-read when moving across data cells, even though it’s shown in the text area in the screenshot… maybe because pseudo elements???
Screenshot 2021-02-09 at 16 50 33

Heydon’s very good inclusive components has the same problem we have, moving across the data cells re-reads the “sort by…” https://inclusive-components.design/data-tables/
Screenshot 2021-02-09 at 16 52 33

My previous decision was bad, it’s better to reliably read the heading with unhelpful “sort by…”  prompt when on a data cell than not read the heading. If we can’t find an ideal solution we should do that?

**Comment by chee** [comment]
@notlee sorry, I wasn't clear when i was saying i couldn't see a difference. What I"m saying is that in my experiments, the table we already have seems to work fine in Orca, NVDA and JAWS!

**Comment by chee** [comment]
> @notlee sorry, I wasn't clear when i was saying i couldn't see a difference. What I"m saying is that in my experiments, the table we already have seems to work fine in Orca, NVDA and JAWS!

that is to say: as long as you go deep enough to select the button, they all read the button title and do not read it when you move between columns

**Comment by notlee** [comment]
@chee oh!! so there isn't an issue with the current implementation for each of your test scenarios? nice 😄

**Comment by origamiserviceuser** [comment]
o-table bundle size difference from 8.1.4 to bee8e27f9006cfa0a4fe22eeb2c131816a055c66
js: 0.26kb increase (0.10kb/gzip)
An insignificant difference was also found for: css (master), css (internal), css (whitelabel)

**Comment by origamiserviceuser** [comment]
o-table bundle size difference from 8.1.4 to 53d04032bdf6ebf38c07b372df43bc923493ea3a
js: 0.26kb increase (0.10kb/gzip)
An insignificant difference was also found for: css (master), css (internal), css (whitelabel)

**Pull request closed without merging**